Rubberduck.CodeAnalysis/Inspections/Concrete/IllegalAnnotationInspection.cs

```diff
@@ -1,175 +1,64 @@
-            {
-                var name = Identifier.GetName(context.annotationName().unrestrictedIdentifier());
-                var annotationType = (AnnotationType) Enum.Parse(typeof (AnnotationType), name, true);
-
-                var moduleHasMembers = _members.Value.Any();
-
-                var isMemberAnnotation = annotationType.HasFlag(AnnotationType.MemberAnnotation);
-                var isModuleAnnotation = annotationType.HasFlag(AnnotationType.ModuleAnnotation);
-
-                var isModuleAnnotatedForMemberAnnotation = isMemberAnnotation
-                    && (_currentScopeDeclaration?.DeclarationType.HasFlag(DeclarationType.Module) ?? false);
+            return illegalAnnotations.Select(annotation => 
+                new QualifiedContextInspectionResult(
+                    this, 
+                    string.Format(InspectionResults.IllegalAnnotationInspection, annotation.Context.annotationName().GetText()), 
+                    new QualifiedContext(annotation.QualifiedSelection.QualifiedName, annotation.Context)));
+        }
 
-                var isMemberAnnotatedForModuleAnnotation = isModuleAnnotation 
-                    && (_currentScopeDeclaration?.DeclarationType.HasFlag(DeclarationType.Member) ?? false);
+        private ICollection<IAnnotation> UnboundAnnotations(IEnumerable<IAnnotation> annotations, IEnumerable<Declaration> userDeclarations)
+        {
+            var boundAnnotations = userDeclarations.SelectMany(declaration => declaration.Annotations)
+                .ToDictionary(annotation => annotation.QualifiedSelection);
+            
+            return annotations.Where(annotation => !boundAnnotations.ContainsKey(annotation.QualifiedSelection)).ToList();
+        }
 
-                var isIllegal = !(isMemberAnnotation && moduleHasMembers && !_isFirstMemberProcessed) &&
-                                (isMemberAnnotatedForModuleAnnotation || isModuleAnnotatedForMemberAnnotation);
+        private ICollection<IAnnotation> NonModuleAnnotationsOnModules(IEnumerable<Declaration> userDeclarations)
+        {
+            return userDeclarations
+                .Where(declaration => declaration.DeclarationType.HasFlag(DeclarationType.Module))
+                .SelectMany(moduleDeclaration => moduleDeclaration.Annotations)
+                .Where(annotation => !annotation.AnnotationType.HasFlag(AnnotationType.ModuleAnnotation))
+                .ToList();
+        }
 
-                if (isIllegal)
-                {
-                    _contexts.Add(new QualifiedContext<ParserRuleContext>(CurrentModuleName, context));
-                }
-            }
+        private ICollection<IAnnotation> NonMemberAnnotationsOnMembers(IEnumerable<Declaration> userDeclarations)
+        {
+            return userDeclarations
+                .Where(declaration => !declaration.DeclarationType.HasFlag(DeclarationType.Module) && declaration.DeclarationType != DeclarationType.Project)
```

Comment:
Won't `.Where(declaration => declaration.DeclarationType.HasFlag(DeclarationType.Member)` do the same thing?

Reply:
No, it does not: e.g. variables and enumeration members do not have the `Member` flag.

Reply:
Does that mean that a variable is not a member but can have a member annotation? If yes, that's misleading.